// @symphony/automation/index.mjs
// @symphony/automation — SDK shim that talks to the running symphony daemon.
//
// All side effects flow through the daemon via authenticated HTTP so dry-run,
// dedup, and audit live in one place. The script author never branches on
// dryRun: the SDK transparently turns mutating calls into would-do reports.

// Env-var lookup is deferred so describe-mode (where the runner imports the
// file just to read its metadata) doesn't fail on a cold module load.

function runtime() {
  const base = process.env.MERIDIAN_AUTOMATION_BASE;
  const token = process.env.MERIDIAN_AUTOMATION_TOKEN;
  const runId = process.env.MERIDIAN_AUTOMATION_RUN_ID;
  if (!base || !token || !runId) {
    throw new Error(
      "@symphony/automation SDK called outside the symphony runner — " +
        "env vars MERIDIAN_AUTOMATION_BASE/TOKEN/RUN_ID missing"
    );
  }
  return {
    base,
    token,
    runId,
    dry: process.env.MERIDIAN_AUTOMATION_DRY === "1",
  };
}

async function call(path, payload) {
  const rt = runtime();
  const res = await fetch(`${rt.base}${path}`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-automation-token": rt.token,
    },
    body: JSON.stringify({ runId: rt.runId, dryRun: rt.dry, ...payload }),
  });
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`SDK ${path} ${res.status}: ${text || res.statusText}`);
  }
  return res.json();
}

function reviveDates(item) {
  if (item && typeof item.updatedAt === "string") {
    item.updatedAt = new Date(item.updatedAt);
  }
  return item;
}

function serializeFilter(filter) {
  const f = { ...(filter || {}) };
  if (f.updatedSince instanceof Date) {
    f.updatedSince = f.updatedSince.toISOString();
  }
  return f;
}

export function defineAutomation(def) {
  if (!def || typeof def !== "object") {
    throw new TypeError("defineAutomation expects an object");
  }
  if (typeof def.name !== "string" || !def.name.trim()) {
    throw new TypeError("defineAutomation: name is required");
  }
  if (!def.schedule || typeof def.schedule !== "object") {
    throw new TypeError("defineAutomation: schedule is required");
  }
  if (typeof def.run !== "function") {
    throw new TypeError("defineAutomation: run() is required");
  }
  return def;
}

export const symphony = {
  github: {
    async issues(filter) {
      const r = await call("/github/issues", { filter: serializeFilter(filter) });
      return (r.items || []).map(reviveDates);
    },
    async prs(filter) {
      const r = await call("/github/prs", { filter: serializeFilter(filter) });
      return (r.items || []).map(reviveDates);
    },
  },
  inbox: {
    async create(entry) {
      if (!entry || typeof entry !== "object") {
        throw new TypeError("inbox.create expects an entry object");
      }
      if (typeof entry.title !== "string" || !entry.title.trim()) {
        throw new TypeError("inbox.create: title is required");
      }
      if (typeof entry.dedupKey !== "string" || !entry.dedupKey.trim()) {
        throw new TypeError("inbox.create: dedupKey is required");
      }
      return call("/inbox/create", { entry });
    },
  },
  tabs: {
    async open(tab) {
      if (!tab || typeof tab !== "object") {
        throw new TypeError("tabs.open expects a tab object");
      }
      if (typeof tab.url !== "string" || !tab.url) {
        throw new TypeError("tabs.open: url is required");
      }
      if (typeof tab.dedupKey !== "string" || !tab.dedupKey.trim()) {
        throw new TypeError("tabs.open: dedupKey is required");
      }
      return call("/tabs/open", { tab });
    },
  },
};

export default { defineAutomation, symphony };
